{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "initial_id",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import pyarrow as pa\n",
    "import pyarrow.parquet as pq\n",
    "import time\n",
    "import os\n",
    "from scipy.optimize import root_scalar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e5cb5efb-b902-4001-b81d-4c01253adb82",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# A function to compute minor allele frequency\n",
    "def eaf_to_maf(eaf):\n",
    "    if eaf<=0.5:\n",
    "        return eaf\n",
    "    else:\n",
    "        return 1-eaf\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17804e44-eeda-4589-9d7e-fee31703625e",
   "metadata": {},
   "source": [
    "# </br> Merging metabolite_1 and metabolite_2 summary statistics files:<br>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "83754b44-df8d-4d98-8372-d92dce73a04e",
   "metadata": {},
   "outputs": [],
   "source": [
    "met2_file=\"../metabolite_sumstats/GCST90200425_harmonized.parquet.gz\"\n",
    "met1_file=\"../metabolite_sumstats/GCST90199752_harmonized.parquet.gz\"\n",
    "\n",
    "\n",
    "met1 = pd.read_parquet(met1_file)\n",
    "met2 = pd.read_parquet(met2_file)\n",
    "gwas_catalog_id_1 = met1_file.split('/')[-1].split('.')[0].split('_')[0]\n",
    "gwas_catalog_id_2 = met2_file.split('/')[-1].split('.')[0].split('_')[0]\n",
    "# Reading metabolite ratio sum stats file\n",
    "\n",
    "ratio_sumstat = pd.merge(met1, met2, suffixes=['_1', '_2'], how='inner',\n",
    "                           on=['position', 'chromosome', 'effect_allele', 'reference_allele', 'pos_name',\n",
    "                               'variant_id'\n",
    "                               ])  # Merging with the ratio on position of SNPs\n",
    "maf_threshold=0.01\n",
    "\n",
    "#        pval_threshold=1e-5\n",
    "ratio_sumstat['maf'] = ratio_sumstat['effect_allele_frequency_1'].apply(eaf_to_maf)\n",
    "ratio_sumstat = ratio_sumstat[ratio_sumstat.maf > maf_threshold]\n",
    "\n",
    "ratio_sumstat['z_1'] = ratio_sumstat['beta_1'] / ratio_sumstat['se_1']\n",
    "ratio_sumstat['z_2'] = ratio_sumstat['beta_2'] / ratio_sumstat['se_2']\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a30b2736-0250-4469-a270-44095c858538",
   "metadata": {},
   "source": [
    "# </br>Obtaining The Correlation Term:<br>\n",
    "## Choosing A correlation term that would make the lambda (genomic inflation factor) equal to 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28d61181-8621-4083-8981-33acef939c8e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "13becdf0-7d04-4063-9a4f-a99bf4c9849c",
   "metadata": {},
   "source": [
    "</br>ratio_sumstat should have the following columns at least: <br>\n",
    "</br>beta_1 : effect size of the numerator summary statistics file\n",
    "</br>beta_2 : effect size of the denominator summary statistics file\n",
    "</br>se_1   : standard error of the numerator summary statistics file\n",
    "</br>se_2   : standard error of the denominator summary statistics file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b5d6cb8-82f0-44f3-8d73-d330215569de",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c9c10f61-d32e-4ead-bea7-b3a608d1ff14",
   "metadata": {},
   "outputs": [],
   "source": [
    "def objective(correlation, df):\n",
    "    \"\"\"Objective function for root_scalar, now accepting a dataframe.\"\"\"\n",
    "    beta_built_ratio = df['beta_1'] - df['beta_2']\n",
    "    se_built_ratio = df['se_1']**2 + df['se_2']**2 - 2 * correlation * df['se_1'] * df['se_2']\n",
    "    se_built_ratio[se_built_ratio < 0] = 0 # Handle potential floating point errors\n",
    "    se_built_ratio = np.sqrt(se_built_ratio)\n",
    "    \n",
    "    z_built_ratio = np.divide(beta_built_ratio, se_built_ratio, \n",
    "                              out=np.zeros_like(beta_built_ratio), \n",
    "                              where=se_built_ratio!=0)\n",
    "\n",
    "    lamb = np.median(z_built_ratio**2) / 0.454936423119572\n",
    "    return lamb - 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "fba4638b-18dc-471d-9afa-b6045a99f78e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "result = root_scalar(objective, args=(ratio_sumstat,), bracket=[-1, 1], method='brentq')\n",
    "correlation = result.root\n",
    "\n",
    "\n",
    "ratio_sumstat['z_1'] = ratio_sumstat['beta_1'] / ratio_sumstat['se_1']\n",
    "ratio_sumstat['z_2'] = ratio_sumstat['beta_2'] / ratio_sumstat['se_2']\n",
    "\n",
    "ratio_sumstat = ratio_sumstat.loc[ratio_sumstat.beta_1 * ratio_sumstat.beta_2 < 0]\n",
    "\n",
    "se_calc = np.sqrt(ratio_sumstat['se_1']**2 + ratio_sumstat['se_2']**2 - 2 * correlation * ratio_sumstat['se_1'] * ratio_sumstat['se_2'])\n",
    "ratio_sumstat['se'] = se_calc\n",
    "ratio_sumstat['beta'] = ratio_sumstat['beta_1'] - ratio_sumstat['beta_2']\n",
    "ratio_sumstat['z'] = ratio_sumstat['beta'] / ratio_sumstat['se']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "263c8290-23fd-4418-a37f-e9673aa36dd7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
